lib/internal/modules/cjs/loader.js

```diff
@@ -1529,6 +1560,113 @@ function getMaybeCachedSource(mod, filename) {
+    setOwnProperty(process, 'mainModule', undefined);
+  } else {
+    const {
+      wrap,
+      namespace,
+    } = cascadedLoader.importSyncForRequire(mod, filename, content, isMain, mod[kModuleParent]);
+    if (!ObjectHasOwn(namespace, 'default') || ObjectHasOwn(namespace, '__esModule')) {
+      mod.exports = namespace;
+    } else {
+      mod.exports = createRequiredModuleFacade(wrap);
+    }
+  }
+}
+
+function loadCTS(module, filename) {
+  if (StringPrototypeIncludes(filename, 'node_modules')) {
+    throw new ERR_UNSUPPORTED_NODE_MODULES_TYPE_STRIPPING(filename);
+  }
+  const source = getMaybeCachedSource(module, filename);
+  const { tsParse } = require('internal/modules/helpers');
+  const content = tsParse(source);
+  module._compile(content, filename, 'commonjs');
+}
+
+/**
+ * Built-in handler for `.ts` files.
+ * @param {Module} module The module to compile
+ * @param {string} filename The file path of the module
+ */
+function loadTS(module, filename) {
+  if (StringPrototypeIncludes(filename, 'node_modules')) {
+    throw new ERR_UNSUPPORTED_NODE_MODULES_TYPE_STRIPPING(filename);
+  }
+
+  // If already analyzed the source, then it will be cached.
+  const source = getMaybeCachedSource(module, filename);
+  const { tsParse } = require('internal/modules/helpers');
+  const content = tsParse(source);
+  let format;
+  if (StringPrototypeEndsWith(filename, '.ts')) {
+    const pkg = packageJsonReader.getNearestParentPackageJSON(filename);
+    // Function require shouldn't be used in ES modules.
+    if (pkg?.data.type === 'module') {
+      if (getOptionValue('--experimental-require-module')) {
+        module._compile(content, filename, 'module');
+        return;
+      }
+
+      const parent = module[kModuleParent];
+      const parentPath = parent?.filename;
+      const packageJsonPath = path.resolve(pkg.path, 'package.json');
+      const usesEsm = containsModuleSyntax(content, filename);
+      const err = new ERR_REQUIRE_ESM(filename, usesEsm, parentPath,
+                                      packageJsonPath);
+      // Attempt to reconstruct the parent require frame.
+      if (Module._cache[parentPath]) {
+        let parentSource;
+        try {
+          parentSource = tsParse(fs.readFileSync(parentPath, 'utf8'));
+        } catch {
+          // Continue regardless of error.
+        }
+        if (parentSource) {
+          const errLine = StringPrototypeSplit(
+            StringPrototypeSlice(err.stack, StringPrototypeIndexOf(
+              err.stack, '    at ')), '\n', 1)[0];
+          const { 1: line, 2: col } =
+              RegExpPrototypeExec(/(\d+):(\d+)\)/, errLine) || [];
+          if (line && col) {
+            const srcLine = StringPrototypeSplit(parentSource, '\n')[line - 1];
+            const frame = `${parentPath}:${line}\n${srcLine}\n${
+              StringPrototypeRepeat(' ', col - 1)}^\n`;
+            setArrowMessage(err, frame);
+          }
```

Comment:
This seems copy-pasted from `Module._extensions['.js']`. I think we should avoid copy-pasting code in here, especially long and complicated blocks like this. Can we create helper functions to hold whatever needs to be shared?

Reply:
the part that looks complicated I put it in a separate function